### PR TITLE
Fast node

### PR DESCRIPTION
**Pull request by kyrie-yl**
### Description

add a description of your changes here...

### Rationale

tell us why we need these changes...

### Example

add an example CLI or API response...

### Changes

Notable changes: 
* add each change in a bullet point here
* ...

### CODE REVIEW

**Comment by guagualvcha** [comment]
The PR description && title of PR is not up to standard. Please give more background and details.

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -122,6 +122,14 @@ var (
 		Name:  "disablesnapprotocol",
 		Usage: "Disable snap protocol",
 	}
+	DisableDiffProtocolFlag = cli.BoolFlag{
+		Name: "disablediffprotocol",
+		Usage: "Disable diff protocol",
```
Comment (guagualvcha) [inline cmd/utils/flags.go:127]: Please add more description.  When should user enable or disable it , what is the default option.

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -122,6 +122,14 @@ var (
...
+	}
+	EnableTrustProtocolFlag = cli.BoolFlag{
+		Name: "enabletrustprotocol",
+		Usage: "Enable trust protocol",
```
Comment (guagualvcha) [inline cmd/utils/flags.go:131]: so be here.

File: cmd/utils/flags.go [hunk 3]
```diff
@@ -264,9 +272,11 @@ var (
 		Usage: "The layer of tries trees that keep in memory",
 		Value: 128,
 	}
-	AllowInsecureNoTriesFlag = cli.BoolTFlag{
-		Name:  "allow-insecure-no-tries",
-		Usage: `Disable the tries state root verification, the state consistency is no longer 100% guaranteed, diffsync is not allowed if enabled. Do not enable it unless you know exactly what the consequence it will cause.`,
+	defaultVerifyMode = ethconfig.Defaults.TriesVerifyMode
+	TriesVerifyModeFlag = TextMarshalerFlag{
+		Name: "tries-verify-mode",
+		Usage: `tries verify mode: "local", "full", "insecure", "none"`,
```
Comment (guagualvcha) [inline cmd/utils/flags.go:278]: More details about the meaning of each mode, how can user choose.

File: core/block_validator.go [hunk 4]
```diff
@@ -34,18 +35,24 @@ const badBlockCacheExpire = 30 * time.Second
...
 // NewBlockValidator returns a new block validator which is safe for re-use
-func NewBlockValidator(config *params.ChainConfig, blockchain *BlockChain, engine consensus.Engine) *BlockValidator {
+func NewBlockValidator(config *params.ChainConfig, blockchain *BlockChain, engine consensus.Engine, mode VerifyMode, peers verifyPeers) *BlockValidator {
 	validator := &BlockValidator{
 		config: config,
 		engine: engine,
 		bc:     blockchain,
 	}
+	if mode.NeedRemoteVerify() {
+		log.Info("this node is a fast node with remote state verifier.")
```
Comment (guagualvcha) [inline core/block_validator.go:52]: Kindly remind: I think this is debug log, please remove it before this PR been merged

File: core/remote_state_verifier.go [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+	// Subscription
+	chainHeadCh  chan ChainHeadEvent
+	chainHeadSub event.Subscription
+
+	// Channels
+	verifyCh  chan common.Hash
+	messageCh chan verifyMessage
+}
+
+func NewVerifyManager(blockchain *BlockChain, peers verifyPeers, allowUntrusted bool) *remoteVerifyManager {
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:56]: `allowUntrusted` ==> `allowInsecure`

File: core/remote_state_verifier.go [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+	// find header of H-11 block.
+	header = vm.bc.GetHeaderByNumber(header.Number.Uint64() - maxForkHeight)
+	// If start from genesis block, there has not a H-11 block.
+	if header == nil {
+		return true
+	}
+	// check whether H-11 block is a empty block.
+	if header.TxHash == types.EmptyRootHash {
+		parent := vm.bc.GetHeaderByHash(header.ParentHash)
+		return header.Root == parent.Root
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:176]: Please handle if `parent` is nil

File: core/remote_state_verifier.go [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+	"fmt"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/metrics"
+	"math/rand"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:13]: Please re-order the import packagees.

File: core/remote_state_verifier.go [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+	// corresponding block is larger than it, the task should be pruned.
+	pruneHeightDiff = 15
+	pruneInterval   = 5 * time.Second
+	resendInterval  = 2 * time.Second
+	// tryAllPeersTime is the time that a block has not been verified and then try all the valid verify peers.
+	tryAllPeersTime = 15 * time.Second
+)
+
+var (
+	remoteVerifyTaskCounter      = metrics.NewRegisteredCounter("remote/state/verify/task/total", nil)
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:33]: The layout of these metrics are not proper, please fix them.

File: core/remote_state_verifier.go [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+				return
+			}
+			diffLayer := vm.bc.GetTrustedDiffLayer(hash)
+			// if this block has no diff, there is no need to verify it.
+			var err error
+			if diffLayer == nil {
+				if diffLayer, err = vm.bc.GenerateDiffLayer(hash); err != nil {
+					log.Error("failed to get diff layer", "block", hash, "number", header.Number, "error", err)
+					return
+				} else {
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:140]: Why return here?

Comment (guagualvcha) [reply core/remote_state_verifier.go:140]: I guess it should be `else if diffLayer == nil`

File: core/remote_state_verifier.go [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+type verifyMessage struct {
+	verifyResult *VerifyResult
+	peerId       string
+}
+
+type verifyTask struct {
+	diffhash       common.Hash
+	blockHeader    *types.Header
+	candidatePeers verifyPeers
+	BadPeers       map[string]struct{}
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:204]: rename `BadPeers` to `badPeers`

File: core/remote_state_verifier.go [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+	startAt        time.Time
+	allowUntrusted bool
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:206]: `allowUntrusted` ==> `allowInsecure`.

File: core/remote_state_verifier.go [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+func (vt *verifyTask) Start(verifyCh chan common.Hash) {
+	vt.startAt = time.Now()
+
+	vt.sendVerifyRequest(defaultPeerNumber)
+	resend := time.NewTicker(resendInterval)
+	defer resend.Stop()
+	for {
+		select {
+		case msg := <-vt.messageCh:
+			switch msg.verifyResult.Status.Code / 100 {
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:235]: Calculation on Status is not allowed.

File: core/remote_state_verifier.go [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+						vt.compareRootHashAndWrite(msg, verifyCh)
+					}
+					newRecievedMsgTypeGauge("partialVerified", msg.peerId).Inc(1)
+				}
+
+			case 2, 4:
+				vt.BadPeers[msg.peerId] = struct{}{}
+				log.Info("peer %s is not available: code %d, msg %s,", msg.peerId, msg.verifyResult.Status.Code, msg.verifyResult.Status.Msg)
+				switch msg.verifyResult.Status {
+				case types.StatusDiffHashMismatch:
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:253]: Too many duplicated code, please use map to reduce it.

File: core/remote_state_verifier.go [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+	var validPeers []VerifyPeer
+	candidatePeers := vt.candidatePeers.GetVerifyPeers()
+	for _, p := range candidatePeers {
+		if _, ok := vt.BadPeers[p.ID()]; !ok {
+			validPeers = append(validPeers, p)
+		}
+	}
+	// if has not valid peer, log warning.
+	if len(validPeers) == 0 {
+		log.Warn("there is no valid peer for block", vt.blockHeader.Number)
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:296]: It should return here.

File: core/remote_state_verifier.go [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+	}
+	if n < 0 || n >= len(validPeers) {
+		for _, p := range validPeers {
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:299]: duplicated code

Comment (guagualvcha) [reply core/remote_state_verifier.go:299]: What about:

```

	if n < len(validPeers) && n > 0 {
		rand.Seed(time.Now().UnixNano())
		rand.Shuffle(len(validPeers), func(i, j int) { validPeers[i], validPeers[j] = validPeers[j], validPeers[i] })
	} else {
		n = len(validPeers)
	}

	for i := 0; i < n; i++ {
		p := validPeers[i]
		p.RequestRoot(vt.blockHeader.Number.Uint64(), vt.blockHeader.Hash(), vt.diffhash)
	}
```

File: core/remote_state_verifier.go [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+	// if n < len(validPeers), select n peers from validPeers randomly.
+	rand.Seed(time.Now().UnixNano())
+	rand.Shuffle(len(validPeers), func(i, j int) { validPeers[i], validPeers[j] = validPeers[j], validPeers[i] })
+	for i := 0; i < n; i++ {
+		p := validPeers[i]
+		p.RequestRoot(vt.blockHeader.Number.Uint64(), vt.blockHeader.Hash(), vt.diffhash)
+	}
+}
+
+func (vt *verifyTask) compareRootHashAndWrite(msg verifyMessage, verifyCh chan common.Hash) {
```
Comment (guagualvcha) [inline core/remote_state_verifier.go:314]: There is no `write` here, please change `compareRootHashAndWrite` to `compareRootHashAndMark`

File: core/remote_state_verifier.go (hunk 5 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:56]: fixed

File: core/remote_state_verifier.go (hunk 7 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:13]: fixed

File: core/remote_state_verifier.go (hunk 6 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:176]: fixed

File: core/remote_state_verifier.go (hunk 8 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:33]: fixed

File: core/remote_state_verifier.go (hunk 9 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:140]: fixed

File: core/remote_state_verifier.go (hunk 10 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:204]: fixed

File: core/remote_state_verifier.go (hunk 11 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:206]: fixed

File: core/remote_state_verifier.go (hunk 12 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:235]: fixed

File: core/remote_state_verifier.go (hunk 14 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:296]: fixed

File: core/remote_state_verifier.go (hunk 13 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:253]: fixed

File: core/remote_state_verifier.go (hunk 15 above)
Comment (kyrie-yl) [reply core/remote_state_verifier.go:299]: fixed

File: ethclient/ethclient.go [hunk 17]
```diff
@@ -22,6 +22,7 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"github.com/ethereum/go-ethereum/core"
```
Comment (guagualvcha) [inline ethclient/ethclient.go:25]: Import order

**Pull request closed without merging**